[SelectiveMasking] Use QUuid to reference symbol layer when masking

troopa81 opened the pull request:
Fixes #34352

This PR is a follow up and implementation of this [comment](https://github.com/qgis/QGIS/pull/47459#issuecomment-1123018070).

Selective masking was using a tuple *symbol key (rule for instance) / symbol layer index path* to reference symbol layer. It's very fragile and is a problem when it comes to cloning symbol layer for instance. This PR propose to replace them with a generated QUuid.

On the API side of things I had to add some *V2* methods in order to not break the API. I changed only the return type of QgsVectorLayerUtils::labelMasks and QgsVectorLayerUtils::symbolLayerMasks but there are not part of Python API.

I implemented a migration method to port from old project. I set the migration code as deprecated in order to remove it when we switch to QGIS 4, so it won't be possible to migrate QGIS project which version is less than 3.28 in QGIS 4. Is it OK?

troopa81 committed:
[Selective masking]Fixes #34352 : use QUUid to identify symbol layer

Don't use pointers anymore because there are not stable when it comes
to clone

troopa81 committed:
Fix vector rendering of fill symbol layer

troopa81 committed:
open report automatically if errors

troopa81 committed:
Fix random generator fill

troopa81 committed:
rename qgs project path

troopa81 committed:
fix project migration of old reference to new ones

troopa81 committed:
fix tests

troopa81 committed:
fix documentation

troopa81 committed:
don't deprecate a class, sip fails

nyalldawson commented:
Nice work @troopa81 -- this should fix a lot of things!

One consideration we need to make is when a clone operation is used to "copy" (or duplicate) a symbol (or a renderer, and accordingly all its symbols).

Eg if I right click a layer and duplicate, then the new layer will need to regenerate all the symbol layer ids. Similarly if I were to press the "duplicate symbol layer" button here then the new layer will need a new id:

![image](https://user-images.githubusercontent.com/1829991/199846630-0a1e8f9c-50b2-4cf2-9204-e209563dbf57.png)

And lastly -- I think there's some places where we should **never** store symbol layer ids. E.g. when we add a symbol to a QgsStyle library, we'll want to clear out all the ids so that reusing the stored symbols doesn't re-use their original ids.  (For the same reason any of the "copy symbol" actions should clear the stored ids too!)

troopa81 committed:
fix text renderer tests

troopa81 committed:
fix api doc

troopa81 committed:
update project to be saved with 3.29

troopa81 committed:
fix duplicate layer

troopa81 committed:
fix doc and typos

troopa81 committed:
Fix QgsStyle addSymbol

troopa81 commented:
> Eg if I right click a layer and duplicate, then the new layer will need to regenerate all the symbol layer ids.

Good catch, fixed! 

Regarding the copy/paste style from a layer to another (or saving/loading to/from a qml file), I don't think it's that bad for the id to be copied as long as they are unique for a given vector layer. 

> And lastly -- I think there's some places where we should never store symbol layer ids. E.g. when we add a symbol to a QgsStyle library, we'll want to clear out all the ids so that reusing the stored symbols doesn't re-use their original ids. (For the same reason any of the "copy symbol" actions should clear the stored ids too!)

I fixed [here](https://github.com/qgis/QGIS/pull/50786/files#diff-5ca9ad9f5b6f80d49659a95c002977fd9eab698c8acfab060fce03f37e2e63a6R161) and  [there](https://github.com/qgis/QGIS/pull/50786/files#diff-5ca9ad9f5b6f80d49659a95c002977fd9eab698c8acfab060fce03f37e2e63a6R147) in QgsTextBackgroundSettings and in [QgsStyle::addSymbol](https://github.com/qgis/QGIS/pull/50786/files#diff-b80a4a7571b0ed0ed0a818e1680fcc132f3a3b3d5bdf66c2a608baf7d44d90f5R298). Do you see any feature/part of code that could be impacted ?

troopa81 committed:
fix typo

nyalldawson commented:
I will as soon as possible -- my workload is extreme right now, so I can't make any promises as to when this will be.

nyalldawson commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -693,10 +693,8 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
-        {
-          slRefs.insert( QgsSymbolLayerReference( mit.key(), *slIt ) );
-        }
+        for ( QString symbolLayerId : mit.value().symbolLayerIds )
```
```suggestion
        for ( const QString &symbolLayerId : mit.value().symbolLayerIds )
```

nyalldawson commented on src/core/maprenderer/qgsmaprendererjob.cpp:
```diff
@@ -826,7 +824,7 @@ std::vector< LayerRenderJob > QgsMapRendererJob::prepareSecondPassJobs( std::vec
 
     QList<MaskSource> &sourceList = it->second;
-    const QSet<QgsSymbolLayerId> &symbolList = it->first;
+    const QSet<QString> &symbolList = it->first;
```
```suggestion
    const QSet<QString> symbolList = it->first;
```

nyalldawson commented on src/core/qgsrendercontext.h:
```diff
@@ -143,6 +143,31 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
+     * \since QGIS 3.12
+     * \deprecated since QGIS 3.30 and replaced with setDisabledSymbolLayersV2
+     */
+    Q_DECL_DEPRECATED void setDisabledSymbolLayers( const QSet<const QgsSymbolLayer *> &symbolLayers );
```
```suggestion
    Q_DECL_DEPRECATED void setDisabledSymbolLayers( const QSet<const QgsSymbolLayer *> &symbolLayers ) SIP_DEPRECATED;
```

nyalldawson commented on src/core/qgsrendercontext.h:
```diff
@@ -151,8 +176,20 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
      * \since QGIS 3.12
+     * \deprecated since QGIS 3.30 and replaced with disabledSymbolLayersV2
+     */
+    Q_DECL_DEPRECATED QSet<const QgsSymbolLayer *> disabledSymbolLayers() const;
```
```suggestion
    Q_DECL_DEPRECATED QSet<const QgsSymbolLayer *> disabledSymbolLayers() const SIP_DEPRECATED;
```

nyalldawson commented on src/core/qgsrendercontext.h:
```diff
@@ -151,8 +176,20 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
+     * \since QGIS 3.30
      */
-    QSet<const QgsSymbolLayer *> disabledSymbolLayers() const { return mDisabledSymbolLayers; }
+    const QSet<QString> &disabledSymbolLayersV2() const;
```
```suggestion
    QSet<QString> disabledSymbolLayersV2() const;
```

Since QSet is implicitly shared, there's no advantage in return a const ref (and doing so causes potential issues in sip/python code)

nyalldawson commented on tests/src/python/test_selective_masking.py:
```diff
@@ -1095,13 +1191,130 @@ def test_layout_export_2_sources_masking(self):
+
+        map_settings.setLayers([layer])
+
+        self.check_layout_export("layout_export_line_pattern_fill", 0, [layer], extent=extent)
```
The text in this image is very small and non-bold, so this test won't be stable across platforms. Can you make it bigger and bold, and ensure that it uses the standard QGIS test font too?

troopa81 replied:
Done

nyalldawson commented on tests/src/python/test_selective_masking.py:
```diff
@@ -1095,13 +1191,130 @@ def test_layout_export_2_sources_masking(self):
+
+        map_settings.setLayers([layer])
+
+        self.check_layout_export("layout_export_random_generator_fill", 0, [layer], extent=extent)
```
Make sure these tests use the standard test font

troopa81 replied:
Done

nyalldawson commented on src/core/textrenderer/qgstextmasksettings.h:
```diff
@@ -196,14 +196,14 @@ class CORE_EXPORT QgsTextMaskSettings
+     * \see setMaskedSymbolLayers()
      */
-    QList<QgsSymbolLayerReference> maskedSymbolLayers() const;
+    const QList<QgsSymbolLayerReference> &maskedSymbolLayers() const;
```
```suggestion
    QList<QgsSymbolLayerReference> maskedSymbolLayers() const;
```

(as noted earlier)

nyalldawson commented on src/core/symbology/qgssymbollayer.cpp:
```diff
@@ -930,3 +925,45 @@ void QgsSymbolLayer::prepareMasks( const QgsSymbolRenderContext &context )
+    context.painter()->setClipPath( mClipPath, Qt::IntersectClip );
+  }
+
+  QgsSymbol *lSubSymbol = recursive ? subSymbol() : nullptr;
+  if ( lSubSymbol )
+  {
+    for ( QgsSymbolLayer *sl : lSubSymbol->symbolLayers() )
+      sl->installMasks( context, true );
+  }
```
```suggestion
  if ( QgsSymbol *lSubSymbol = recursive ? subSymbol() : nullptr )
  {
    for ( QgsSymbolLayer *sl : lSubSymbol->symbolLayers() )
      sl->installMasks( context, true );
  }
```

nyalldawson commented on src/core/symbology/qgssymbollayer.cpp:
```diff
@@ -930,3 +925,45 @@ void QgsSymbolLayer::prepareMasks( const QgsSymbolRenderContext &context )
+    context.painter()->restore();
+  }
+
+  QgsSymbol *lSubSymbol = recursive ? subSymbol() : nullptr;
+  if ( lSubSymbol )
```
```suggestion
  if ( QgsSymbol *lSubSymbol = recursive ? subSymbol() : nullptr )
```

nyalldawson commented on src/core/symbology/qgssymbollayer.h:
```diff
@@ -626,10 +626,26 @@ class CORE_EXPORT QgsSymbolLayer
+     * This id is unique in the whole project
+     * \since QGIS 3.30
+     */
+    const QString &id() const;
```
```suggestion
    QString id() const;
```

nyalldawson commented on src/core/symbology/qgssymbollayer.h:
```diff
@@ -626,10 +626,26 @@ class CORE_EXPORT QgsSymbolLayer
 
+    /**
+     * Set symbol layer identifier
+     * This id has to be unique in the whole project
```
```suggestion
     * \warning This ID must be unique within a whole project.
```

nyalldawson commented on src/core/symbology/qgsfillsymbollayer.cpp:
```diff
@@ -2576,14 +2576,16 @@ bool QgsLinePatternFillSymbolLayer::hasDataDefinedProperties() const
-void QgsLinePatternFillSymbolLayer::startFeatureRender( const QgsFeature &, QgsRenderContext & )
+void QgsLinePatternFillSymbolLayer::startFeatureRender( const QgsFeature &, QgsRenderContext &context )
 {
   // deliberately don't pass this on to subsymbol here
```
it looks here like the existing comment applies to "installMasks", whereas it's actually referring to the call to startFeatureRender. Can you tweak the comments or add spacing or something to keep this distinction clear?

troopa81 replied:
I moved install/remove masks after the comment and completed it a bit to remove ambiguity

nyalldawson commented on src/core/symbology/qgssymbollayerreference.cpp:
```diff
@@ -43,36 +35,57 @@ QgsSymbolLayerReferenceList stringToSymbolLayerReferenceList( const QString &str
-  QRegularExpressionMatchIterator partsIt = partsRx.globalMatch( str + ';' );
-
-  while ( partsIt.hasNext() )
+  if ( str.contains( "," ) )
```
```suggestion
  if ( str.contains( ',' ) )
```

nyalldawson commented on src/core/symbology/qgssymbollayerreference.cpp:
```diff
@@ -43,36 +35,57 @@ QgsSymbolLayerReferenceList stringToSymbolLayerReferenceList( const QString &str
+  }
+  else
+  {
+    const QStringList elems = str.split( ";" );
```
```suggestion
    const QStringList elems = str.split( ';' );
```

nyalldawson commented on src/core/symbology/qgssymbollayerutils.h:
```diff
@@ -897,6 +898,39 @@ class CORE_EXPORT QgsSymbolLayerUtils
+     * \since QGIS 3.30
+     * \deprecated since QGIS 3.30
+     */
+    Q_DECL_DEPRECATED static void fixOldSymbolLayerReferences( const QMap<QString, QgsMapLayer *> &mapLayers );
```
Maybe QgsProjectFileTransform is a better place for this function to live?

troopa81 replied:
Yes :+1:

troopa81 committed:
Treat review comments

Co-authored-by: Nyall Dawson <nyall.dawson@gmail.com>

troopa81 committed:
fix test generated file permissions

troopa81 committed:
Fix mask

troopa81 commented:
@nyalldawson I took care of all your comments, thank you for reviewing! 

mingw is failing but I don't think it's related (there is no clear reason of why it fails)

nyalldawson commented:
@troopa81 can you rebase and fix the merge conflicts, then we're good to merge this one :+1:

troopa81 commented:
> @troopa81 can you rebase and fix the merge conflicts, then we're good to merge this one

Done, thanks @nyalldawson

thymaro commented:
Just for my understanding, does this mean that line pattern fills can be masked, now?

troopa81 commented:
> Just for my understanding, does this mean that line pattern fills can be masked, now?

Yes. It's available only in master and will be in 3.32
